ecs-agent/acs/session/attach_instance_eni_responder.go

```diff
@@ -0,0 +1,154 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package session
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/pkg/errors"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/attachmentinfo"
+	apieni "github.com/aws/amazon-ecs-agent/ecs-agent/api/eni"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/status"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+)
+
+const (
+	AttachInstanceENIMessageName = "AttachInstanceNetworkInterfacesMessage"
+)
+
+// attachInstanceENIResponder implements the wsclient.RequestResponder interface for responding
+// to ecsacs.AttachInstanceNetworkInterfacesMessage messages sent by ACS.
+type attachInstanceENIResponder struct {
+	eniHandler ENIHandler
+	respond    wsclient.RespondFunc
+}
+
+// NewAttachInstanceENIResponder returns an instance of the attachInstanceENIResponder struct.
+func NewAttachInstanceENIResponder(eniHandler ENIHandler,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &attachInstanceENIResponder{
+		eniHandler: eniHandler,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*attachInstanceENIResponder) Name() string { return "attach instance ENI responder" }
+
+func (r *attachInstanceENIResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handleAttachMessage
+}
+
+func (r *attachInstanceENIResponder) handleAttachMessage(message *ecsacs.AttachInstanceNetworkInterfacesMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", AttachInstanceENIMessageName))
+	receivedAt := time.Now()
+
+	// Validate fields in the message.
+	if err := validateAttachInstanceNetworkInterfacesMessage(message); err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", AttachInstanceENIMessageName), logger.Fields{
+			field.Error: err,
+		})
+		return
+	}
+
+	// Handle ENIs in the message.
+	messageID := aws.StringValue(message.MessageId)
+	for _, mENI := range message.ElasticNetworkInterfaces {
+		expiresAt := receivedAt.Add(time.Duration(aws.Int64Value(message.WaitTimeoutMs)) * time.Millisecond)
+		go func(eni *ecsacs.ElasticNetworkInterface) {
```

Comment:
In the PR notes we mentioned
> Handle ENI attachment in async (no reason for this to be a blocking operation)

Curious how asynchronous processing adds value here? It looks like `HandleENIAttachment` function currently does little work, returns quickly, and will never get stuck. Since the handling logic is straightforward, I wonder if we want to keep the implementation simple.

We should also remember that the handler already spins off new goroutine to handle each incoming message.

Reply:
Thanks for your comment!

> Curious how asynchronous processing adds value here?

I figured since each ENI's attachment is independent from each other and independent from ACKing receipt of valid message from ACS, it would be a best practice not to block here before allowing other such independent tasks to start. This also follows the convention of other ACS responders such as the heartbeat responder (example [here](https://github.com/aws/amazon-ecs-agent/blob/master/ecs-agent/acs/session/heartbeat_responder.go#L50-L71)).

> It looks like `HandleENIAttachment` function currently does little work, returns quickly, and will never get stuck. Since the handling logic is straightforward, I wonder if we want to keep the implementation simple.

That would actually depend on the implementation of the `HandleENIAttachment` function by the `ENIHandler` we use. The current implementation used by the Docker-based Agent defined in `agent/` directory can be found [here](https://github.com/aws/amazon-ecs-agent/blob/master/agent/acs/handler/attach_eni_handler_common.go#L42-L66). In my opinion, when I do the full code trace of this function, I'm not sure that I would consider the implementation here to do "little work". We should also consider other use cases such as:
1. What if an `ENIHandler` that has a different implementation which does significant work, returns slowly, and/or has potential to get stuck is used?
2. What if the `ENIHandler` that we currently use is changed in the future such that it does significant work, returns slowly, and/or has potential to get stuck?

> We should also remember that the handler already spins off new goroutine to handle each incoming message.

Does the responder spin off a new goroutine to handle each incoming message? Could you point to in where in code you are seeing this? My current understanding under this PR is that is not the case, but I could be missing something.

Reply:
Thanks for the detailed explanation! 

You are absolutely right that we should be future-proofing and that the ENI handling may grow more complex in the future. 

Regarding the top-level handler, apologies that it was a misreading on my part. I was referring to https://github.com/aws/amazon-ecs-agent/blob/master/ecs-agent/wsclient/client.go#L447. Reading again, the code use the same goroutine for consuming all messages, so there doesn't look like to be built-in per-message goroutine.